### DISCUSSION TITLE
EVP_PKEY_CTX_set_signature_md() fails for blake2 and sm3

### DISCUSSION THREAD

**Discussion by HenrikHolst** [openssl/openssl#24562] [Q&A] [answered]
Don't know if I'm doing something really wrong here but when trying to validate a piece of RSA encrypted data with a SM3 or Blake2s256/Blake2b512 signature EVP_PKEY_CTX_set_signature_md() fails with "error:1C8000AE:Provider routines::digest not allowed" but works perfectly fine for other md:s like md5/sha1/sha3-512.

EVP_MD_fetch() loads it just fine so the problem isn't that the digest is missing from my version of OpenSSL (which is v3.0.13 as included in Ubuntu 24.04). They also list just fine under "openssl list -digest-algorithms".

```
/* load the RSA key from a file in PEM format */
EVP_PKEY *rsa;
OSSL_DECODER_CTX *rsa_dctx = OSSL_DECODER_CTX_new_for_pkey (&rsa, "PEM", NULL, "RSA", OSSL_KEYMGMT_SELECT_KEYPAIR, NULL, NULL);
OSSL_DECODER_from_fp (rsa_dctx, f);
OSSL_DECODER_CTX_free (rsa_dctx);

/* set up the signature key */
EVP_PKEY_CTX *rsa_ctx = EVP_PKEY_CTX_new (rsa, NULL);
EVP_PKEY_sign_init (rsa_ctx);
EVP_PKEY_CTX_set_signature_md (rsa_ctx, EVP_MD_fetch (NULL, "SM3", NULL)); /* this call fails with digest not allowed". */

```

**Comment by t8m** [comment, marked as answer]
These digests are not supported for RSA signatures in OpenSSL.

**Reply by levitte** [reply]
> ... are not supported for RSA signatures in *OpenSSL*

Or, for that matter, by [PKCS#1](https://www.rfc-editor.org/rfc/rfc8017#appendix-A.2.4)

**Comment by HenrikHolst** [comment]
Ok so this is a TLS / PKCS thing so one have to perform such a thing manually by first creating the RSA and then signing using the MD interface instead of using the PKEY interface.

**Reply by t8m** [reply]
I do not think using the MD interface (if you mean EVP_DigestSignInit() etc.) would change anything on this matter.

**Reply by HenrikHolst** [reply]
> I do not think using the MD interface (if you mean EVP_DigestSignInit() etc.) would change anything on this matter.

What I mean is that one would create the MD using the MD interface, then apply padding manually and then EVP_PKEY_sign() with no set md function to do a straight RSA signature. Cumbersome when OpenSSL could do it for you but I guess PKCS somehow forbids it so.

**Reply by levitte** [reply]
Sure enough, that's a possible way to do it.

**Comment by HenrikHolst** [comment]
Btw what does EVP_PKEY_sign() do if one never calls EVP_PKEY_CTX_set_signature_md() ? The docs is a bit unclear on this, by doing practical tests it looks like PKCS1 padding is still applied and a RSA signature is produced.

I guess that the main difference is that no OID for the MD is added so while a proper RSA signature is produced so when the signature is to be verified one can only deduct that it is a valid RSA signature for the bytes supplied but not that it was produced by a certain digest function.

**Reply by t8m** [reply]
You have to set the padding to NO padding, if you apply the padding yourself.

**Reply by HenrikHolst** [reply]
> You have to set the padding to NO padding, if you apply the padding yourself.

That I understand, but by default it looks to apply PKCS1 padding (it is what it reports but just wanted to double check since the docs is a bit muddy here).

**Reply by t8m** [reply]
Yes, the PKCS#1 padding is the default.

**Comment by HenrikHolst** [comment]
Sorry to reopen the conversation but I have one more question (hopefully just one).

So I "solved" this for the non supported digests by appending the proper ASN.1 prefix, e.g like this for Whirlpool:

```
unsigned char md2[EVP_MAX_MD_SIZE + 21];
memcpy (md2, "\x30\x4e\x30\x0a\x06\x06\x28\xCF\x06\x03\x00\x37\x05\x00\x04\x40", 16);
memcpy (md2 + 16, md, mdlen);

EVP_PKEY_sign (pctx, sig, &siglen, md2, mdlen + 16);
```

and then doing the exact same on the verify side:
```
unsigned char md2[EVP_MAX_MD_SIZE + 21];
memcpy (md2, "\x30\x4e\x30\x0a\x06\x06\x28\xCF\x06\x03\x00\x37\x05\x00\x04\x40", 16);
memcpy (md2 + 16, md, mdlen);
EVP_PKEY_verify (pctx, rsa_sig_from_wire, rsa_size, md2, mdlen + 16)
```

And this worked fine for Whirlpool, SM3 and blake2s256 but when I tried to do the same with blake2b512 with this ASN.1 prefix:

`memcpy (md2, "\x30\x53\x30\x0f\x06\x0B\x2B\x06\x01\x04\x01\x8D\x3A\x0C\x02\x01\x10\x05\x00\x04\x40", 21);`

Then EVP_PKEY_verify fails with "0200008A:rsa routines::invalid padding". Note that EVP_PKEY_sign() returns 1 so signing works but verify doesn't for this particular case. I'm a bit perplexed that it complains about the padding because AFAIK EVP_PKEY_sign will apply the standard PKCS1 padding (and I checked with EVP_PKEY_CTX_get_rsa_padding and it returns PKCS1 on both the signing and verify contexes).

The total mdlen+21 hists 85 bytes which is uneven but so does blake2s256 (53) and that works, also my RSA key is 4096 bits so 85 should not be too large a chunk to encrypt even with added padding (and as stated EVP_PKEY_sign doesn't complain).

Ofc removing the ASN.1 prefixes makes it also works, but the reason I added them is to have the on-wire protocol compatible with a potential future where some of these are added to the supported digests to be used for RSA (who knows what the future brings).

So am I doing something really dumb here or might there be a issue with EVP_PKEY_verify with not adding the proper padding, or is the padding error simply an indiction of another problem and not the padding as such?
